RFC: `#[cfg(version_since(rust, "1.95"))]` for Rust-version conditional compilation (#3857)
rust-lang/rfcs: text/3857-cfg-version.md

We expect rustc to strip pre-release versions, treating them as if they fully implement that language version.

`rust` will be specified as `--check-cfg 'cfg(rust, values(since("1.95.0")))'`
(or whatever version this gets stabilized in).

This will be reported back through `--print-cfg`.

Because this gets reported back in `--print-cfg`,
Cargo will expose `rust` in:
- build scripts as `CARGO_CFG_RUST`
- `target."cfg()".dependencies`

## clippy

Clippy has a [`clippy::incompatible_msrv`](https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv) lint
which will fire whenever a standard library item is used with a `#[stable(since)]` newer than `package.rust-version`.
However, it will be perfectly reasonable to use those items when guarded by a `#[cfg(since)]`.

Clippy may wish to:
- Find a way to reduce false positives, e.g. evaluating the `cfg(since)`s that led to the item's usage or disabling the lint within `#[cfg(since)]`
- Suggest `#[cfg(since)]` in the `clippy::incompatible_msrv` diagnostic report (maybe along with offering to bump MSRV)

# Drawbacks
[drawbacks]: #drawbacks

> People may be using `--cfg rust` already and would be broken by this change.

Author: At least with my [initial search](https://grep.app/search?f.lang=Rust&f.lang.pattern=rust&case=true&regexp=true&q=cfg.*%5Cbrust%5Cb), I did not see any public uses

and leaves room for related predicates like `before`.
We could also call this `minimum`, or support comparison operators in the spirit of [RFC 3796](https://github.com/rust-lang/rfcs/pull/3796).
The risk with a general word like `since` is if we gain support for other data types in cfgs, like integers for embedded development.
The name `since` might apply in some situations but not others and its unclear if we'd want to generalize it past versions.
While having a specific name avoids these concerns.

We could swap the order of parameters and make `rust` a default for the second parameter to allow `#[cfg(since("1.95"))]` as a shorthand.
However, this would look confusing in Cargo and doesn't seem like its offering enough of a benefit to be worth the costs.

The `ConfigurationVersion` is sloppy with the string literal's syntax (relying on `--check-cfg`) so that
- Allows evolution without requiring an MSRV bump
- Its consistent with other predicates, e.g. `#[cfg(foo = "1.0")]`

If we were stricter on the syntax,
we could allow for version numbers to be directly accepted, without quotes 
(e.g. `#[cfg(since(rust, 1.95.0))]`).

> If we ever decided to support operators (e.g.`#[cfg(since(rust, "=1.95.0"))]`, see `--check-cfg`), then we'd need to decide if those also go outside the string or then require a string, being inconsistent.

Reviewer: This also seems to cause problems when it is combined with the "different versions may have different `cmp` implementations and we may want to account for that by allowing setting a policy" idea, doesn't it?

Then we might be expected to allow for non-SemVer strings in this "bare words" form. Then the parser might have to reason about how to properly consume the tokens instead of just extracting a string, possibly according to a policy which, in the current proposal, would be specified *after* that sequence of tokens? Or we just disallow even parsing those, which seems like it would have other problems.

Author: Included this in 3a7609a4

This would also be inconsistent with other uses of `cfg`s
*but* maybe that would just be the start to natively supporting more types in `cfg`,

> like integers which are of interest to embedded folks.

Reviewer: It feels somewhat unclear to me why we would start this journey by supporting this new "version literal" type in `cfg` only and not in the language? As opposed to supporting integers first, which *are* supported in the language.

- Suggest `#[cfg(since)]` in the `clippy::incompatible_msrv` diagnostic report (maybe along with offering to bump MSRV)

# Drawbacks
[drawbacks]: #drawbacks

People may be using `--cfg rust` already and would be broken by this change.
There are no compatibility concerns with predicate names.

This does not include a solution for adopting this within `Cargo.toml` without waiting for an MSRV bump.

Traditionally, maintainers only test their MSRV and latest, assuming those will catch every issue.
While that isn't always true today (e.g. some Cargo features go from "unknown" warning to "unstable" error to supported and MSRV might be in the warning phase),
having distinct implementations for different Rust versions can make the testing matrix more complex.
Tools like [`cargo hack`](https://crates.io/crates/cargo-hack) can help which can run commands on not just one toolchain version but also the every version starting with the MSRV with a command like `cargo hack --rust-version --version-step 1 check`.

This does not help with probes for specific implementations of nightly features which still requires having a `build.rs`.

> Libraries could having ticking time bombs that accidentally break or have undesired behavior for some future Rust version that can't be found until we hit that version.

Reviewer 2: I wonder if this is something that could be warn-by-default. Or at least warn-by-default in clippy. The warning would only be an issue when a workspace is worked on by rust versions older than their highest version cfg.

> Libraries could having ticking time bombs that accidentally break or have undesired behavior for some future Rust version that can't be found until we hit that version.

Reviewer 3: Also I would assume this problem already exists with `rustversion`.

- This does not conform to SemVer's precedence rules,
  as `-nightly` is an older version than `-beta.4` while [SemVer's precedence rules](https://semver.org/#spec-item-11) say the opposite
- Crater runs and local builds don't necessarily have a version that fits within this picture

## `#[cfg(nightly)]`

Depending on what is meant by this,
we either need the language version or the vendor name and version as well as a way to check for the presence of `pre-release`.

See also [`#[cfg(nightly)]`](https://rust-lang.github.io/rfcs/2523-cfg-path-version.html#cfgnightly) in the previous RFC.

## `cfg_target_version`

Instead of defining a new `#[cfg]` predicate, [RFC 3750](https://github.com/rust-lang/rfcs/pull/3750)
could reuse the `#[cfg(since)]` predicate.

As not all systems use SemVer, we can either
- Contort the version into SemVer
  - This can run into problems either with having more precision (e.g. `120.0.1.10` while SemVer only allows `X.Y.Z`) or post-release versions (e.g. `1.2.0.post1` which a SemVer predicate would treat as a pre-release).
- Add an optional third field for specifying the version format (e.g. `#[cfg(since(windows, "10.0.10240", <policy-name>)]`)

> - Make `--check-cfg` load-bearing by having the version policy name be specified in the `--check-cfg` predicate

Reviewer: One depressingly verbose option, but one that will work without modification to the proposal, is to attach an additional "epoch version" via `cfg`, e.g.

```rust
#[cfg(
    all(
        since(thing, "1.0.0"), since(thing_epoch, "2")
    )
)]
mod something {}

#[cfg(
    all(
        since(thing, "1.1.0"), not(since(thing_epoch, "2"))
    )
)]
mod something {}
```
Obviously, `since(thing, "1.0.0")` would be satisfied by `"1.1.0"` in ordering without the `all` including an additional "version epoch".

Author: Unsure how much we need to explore this as its a future possibility.  The main question is whether we paint ourselves into a corner for `target_version`.

As Cargo mirrors Rust's `#[cfg]` syntax, it too will gain this predicate.

The [syntax](https://doc.rust-lang.org/reference/conditional-compilation.html#grammar-ConfigurationPredicate) is:
```
ConfigurationVersion -> `since` `(` IDENTIFIER `,` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`
```

If `IDENTIFIER` has no value or is undefined, this will evaluate to `false`.

If `IDENTIFIER` is not a valid [SemVer](https://semver.org/) value, this will evaluate to `false`.

If the string literal does not conform to the syntax from `<major>` to `<major>.<minor>.<patch>-<pre-release>` where the first three fields must be integers, this will evaluate to `false`.
Note that this excludes support for the `+build` field.

Otherwise, the `IDENTIFIER` will be compared to the string literal according to
[Cargo's `>=` version requirements](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#comparison-requirements).
For example, `#[cfg(since(rust, "1.90"))]` will be treated as `1.95.2 >= 1.90.0`.

See also `--check-cfg`.

## `--check-cfg`

A new predicate will be added of the form:
```
ConfigurationVersion -> `since` `(` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`

> ```

Reviewer 4: It's unclear to me if we should force a minimum version. We could simply that not providing a minimum supported version means that's it correspond to the lowest SemVer version.

The syntax for the contents of the string literal is a SemVer value without the build field.

> This will specify that the given `--cfg` is valid for all values that match:

Reviewer 4: It's worth noting that `--check-cfg` doesn't check `--cfg` arguments. We do reserve the right to check them though[^1].

[^1]: https://doc.rust-lang.org/nightly/rustc/check-cfg.html#specifying-expected-names-and-values

- SemVer syntax
- from the specified version and up

When the given `--cfg` is used with the `since` predicate:
- the string literal should not be of a syntax that evaluates to `false`
- the minimum version requirement must specify a subset of what the `--check-cfg` specifies

So given `--check-cfg 'cfg(foo, values(since("1.95.0")))'`,

> - ✅ `#[cfg(foo = "1.100.0")]`

Reviewer 4: This should be a warning right? since it doesn't use the `since()` predicate (and would never be evaluated to true anyway?).

```suggestion
- ⚠️ `#[cfg(foo = "1.100.0")]`
```

If someone wants this to not be a warning they should specify that the "1.100.0" string literal value is expected, like so: `cfg(foo, values("1.100.0", since("1.95.0")))`

<!--
Making this well-formed seems undesirable to me. This is will force the compiler to no longer be able to treat simple config (`key = "value"`) as a finite set.
-->

Author: The `check-cfg` `since` predicate is meant to specify a continuous range of `values()`, not limit what operators or predicates can be used.  See also 0d7f46ce.

The trade offs for this are talked about later.

Reviewer 4: So `since()` in `--check-cfg` applies to both simple cfg `key = "value"` and `since(key, "value")`, that seems unexpected to me?

Btw, note that empty `values()` has a specific meaning (*No expected values*) for `--check-cfg`, which I'm assuming is not what you are referring to.

Author: Yes, it applies to both.

And yes, I edited to say `values(...)`.

- ⚠️ `#[cfg(foo = "1.0")]`: not SemVer syntax
- ✅ `#[cfg(since(foo, "1.95.0"))]`
- ✅ `#[cfg(since(foo, "1.100.0"))]`
- ✅ `#[cfg(since(foo, "3.0.0"))]`
- ✅ `#[cfg(since(foo, "1.95"))]`
- ⚠️ `#[cfg(since(foo, "1.90.0"))]`: matches a superset of `--check-cfg`
- ⚠️ `#[cfg(since(foo, "1.95.0-0"))]`: matches a superset of `--check-cfg`
- ⚠️ `#[cfg(since(foo, "1"))]`: matches a superset of `--check-cfg`
- ⚠️ `#[cfg(since(foo, "bar"))]`: invalid string literal syntax

## `rust` cfg

A new built-in cfg `--cfg rust=<version>` will be added by the compiler
that specifies the language version.
This will be the version of `rustc` with the behavior for pre-release versions being unspecified.
We expect rustc to:
- Translate the `-nightly` pre-release to `-incomplete
- Strip the `-beta.5` pre-release

`rust` will be specified as `--check-cfg 'cfg(rust, values(since("1.95.0")))'`
(or whatever version this gets stabilized in).

> This will be reported back through `--print-cfg`.

Reviewer 4: ```suggestion
This will be reported back through `--print=cfg`.
```

> Because this gets reported back in `--print-cfg`,

Reviewer 4: ```suggestion
Because this gets reported back in `--print=cfg`,
```

- Find a way to reduce false positives, e.g. evaluating the `cfg(since)`s that led to the item's usage or disabling the lint within `#[cfg(since)]`
- Suggest `#[cfg(since)]` in the `clippy::incompatible_msrv` diagnostic report (maybe along with offering to bump MSRV as that is a reasonable alternative)

# Drawbacks
[drawbacks]: #drawbacks

People may be using `--cfg rust` already and would be broken by this change.
There are no compatibility concerns with predicate names.

This does not include a solution for adopting this within `Cargo.toml` without waiting for an MSRV bump.

Traditionally, maintainers only test their MSRV and latest, assuming those will catch every issue.
While that isn't always true today (e.g. some Cargo features go from "unknown" warning to "unstable" error to supported and MSRV might be in the warning phase),
having distinct implementations for different Rust versions can make the testing matrix more complex.
Tools like [`cargo hack`](https://crates.io/crates/cargo-hack) can help which can run commands on not just one toolchain version but also the every version starting with the MSRV with a command like `cargo hack --rust-version --version-step 1 check`.

> This does not help with probes for specific implementations of nightly features which still requires having a `build.rs`.

Author: Added more nuance in 545ef525

> This does not help with probes for specific implementations of nightly features which still requires having a `build.rs`.

Author: FYI I mostly reverted that change in 6f656d541455b754d0a1076ce03da20ea19cee1f  as this is encouraging a discussion on what the "right" way to discourage something not even in scope for the RFC (and how much it is right to discourage).

Feel free to unresolve if you have concerns about this.

This may limit us if we decide to allow for alternative version formats like with [target_version](#cfg_target_version) as they may not have formats that map well to SemVer.
Worst case, we'd need to accept arbitrary bare words.
This would also be inconsistent with other uses of `cfg`s
*but* maybe that would just be the start to natively supporting more types in `cfg`,
like integers which are of interest to embedded folks.

## `--check-cfg`

The `--check-cfg` predicate and the value for `rust` ensures users get warnings about
- Invalid syntax
- Using this with versions from before its supported, e.g. `#[cfg(since(rust, "1.0.0")]`

`--check-cfg` requires a SemVer version, rather than a version requirement,
in case we want the future possibility of relaxing SemVer versions
*and* we want to infer from the fields used in `--check-cfg` to specify the maximum number of fields accepted in comparisons.

We could have the `check-cfg` `since` predicate only apply to the `cfg` `since` predicate,

> causing `#[cfg(rust = "1.100.0")]` to warn.

Reviewer 4: I would much prefer that. If someone want's exact matching of version we should IMO have a `#[cfg()]` predicate to do that.

Author: Since we talked, do you still feel this way?

I'm going to be writing up a thing for editions, and I was considering it would look like `--check-cfg "cfg(edition, values("2015", "2018", "2021", "2024", since("2025")))"` to say "the versions are discrete but we don't know the future".  For both `edition` and `rust`, we have reasons to do both exact match and version matches

In addition to the use cases given in the Motivation section,
this will allow users to check for specific nightly versions.

Some challenges for this with `rustc --version`:
- Nightly versions for a given release are mutable,
  all mapping to the `-nightly` pre-release version rather than including the date within the pre-release
- This does not conform to SemVer's precedence rules,
  as `-nightly` is an older version than `-beta.4` while [SemVer's precedence rules](https://semver.org/#spec-item-11) say the opposite
- Crater runs and local builds don't necessarily have a version that fits within this picture

## `#[cfg(nightly)]`

Depending on what is meant by this,
we either need the language version or the vendor name and version as well as a way to check for the presence of `pre-release`.

See also [`#[cfg(nightly)]`](https://rust-lang.github.io/rfcs/2523-cfg-path-version.html#cfgnightly) in the previous RFC.

## `cfg_target_version`

Instead of defining a new `#[cfg]` predicate, [RFC 3750](https://github.com/rust-lang/rfcs/pull/3750)
could reuse the `#[cfg(since)]` predicate.

As not all systems use SemVer, we can either
- Contort the version into SemVer

>   - This can run into problems either with having more precision (e.g. `120.0.1.10` while SemVer only allows `X.Y.Z`) or post-release versions (e.g. `1.2.0.post1` which a SemVer predicate would treat as a pre-release).

Reviewer: This is seemingly self-conflicting: `1.2.0.post1` is actually just invalid SemVer per the first half of the sentence, as SemVer requires prerelease versions to be separated by a hyphen. Perhaps you meant
```suggestion
  - This can run into problems either with having more precision (e.g. `120.0.1.10` while SemVer only allows `X.Y.Z`) or post-release versions (e.g. `1.2.0-post1` which a SemVer predicate would treat as a pre-release).
```

Reviewer: It might be worth noting that "120.0.1-10" would work, and would compare correctly against "120.0.1-11", "120.0.1-9", and such, due to the details of SemVer's prerelease ordering allowing multiple dot-separated identifiers in a prerelease version, and ordering by integer sort if a given identifier is entirely digits.

But you don't need to copy-paste the SemVer docs into this RFC if that information seems extraneous.

Author: I tried clarifying this in 4352a955.

```
ConfigurationVersion -> `since` `(` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`
```

The syntax for the contents of the string literal is a SemVer value without the build field.

This will specify that the given `--cfg` is valid for all values that match:
- SemVer syntax
- from the specified version and up

When the given `--cfg` is used with the `since` predicate:
- the string literal should not be of a syntax that evaluates to `false`
- the minimum version requirement must specify a subset of what the `--check-cfg` specifies

So given `--check-cfg 'cfg(foo, values(since("1.95.0")))'`,
- ✅ `#[cfg(foo = "1.100.0")]`
- ⚠️ `#[cfg(foo = "1.0")]`: not SemVer syntax
- ✅ `#[cfg(since(foo, "1.95.0"))]`
- ✅ `#[cfg(since(foo, "1.100.0"))]`
- ✅ `#[cfg(since(foo, "3.0.0"))]`
- ✅ `#[cfg(since(foo, "1.95"))]`
- ⚠️ `#[cfg(since(foo, "1.90.0"))]`: matches a superset of `--check-cfg`
- ⚠️ `#[cfg(since(foo, "1.95.0-0"))]`: matches a superset of `--check-cfg`
- ⚠️ `#[cfg(since(foo, "1"))]`: matches a superset of `--check-cfg`
- ⚠️ `#[cfg(since(foo, "bar"))]`: invalid string literal syntax

## `rust` cfg

A new built-in cfg `--cfg rust=<version>` will be added by the compiler

> that specifies the language version.

Reviewer 4: Would the `<version>` here be quoted? because if is, then that would be a breaking change as we would start interpreting the inside of the quoted string, potentially breaking user who might do `--cfg my_cfg="126.0"` and expect `#[cfg(my_cfg = "1.127")]` to not match.

Author: This does not change anything about how cfg's are specified.

Traditionally, maintainers only test their MSRV and latest, assuming those will catch every issue.
While that isn't always true today (e.g. some Cargo features go from "unknown" warning to "unstable" error to supported and MSRV might be in the warning phase),
having distinct implementations for different Rust versions can make the testing matrix more complex.
Tools like [`cargo hack`](https://crates.io/crates/cargo-hack) can help which can run commands on not just one toolchain version but also the every version starting with the MSRV with a command like `cargo hack --rust-version --version-step 1 check`.

This does not help with identifying nightlies that a feature is available on or compatible with which will still require a `build.rs`.
In terms of doing this via build probes,
Cargo team has previously rejected support for build probes
([source](https://github.com/rust-lang/cargo/issues/11244#issuecomment-2326780810)).
Whether build probes or nightly version checks,
auto-enabling nightly features
(rather than having users opt-in)
runs counter to the spirit of nightly (works like stable except where you opt-in)

> and cause problems if the checks are incorrect which has broken many crates nightly builds in the past

Reviewer 5: ```suggestion
and cause problems if the checks are incorrect which has broken many crates' nightly builds in the past
```

> This does not help with identifying nightlies that a feature is available on or compatible with which will still require a `build.rs`.
> In terms of doing this via build probes,
> Cargo team has previously rejected support for build probes
> ([source](https://github.com/rust-lang/cargo/issues/11244#issuecomment-2326780810)).
> Whether build probes or nightly version checks,
> auto-enabling nightly features
> (rather than having users opt-in)
> runs counter to the spirit of nightly (works like stable except where you opt-in)
> and cause problems if the checks are incorrect which has broken many crates' nightly builds in the past
> and has even caused friction in changing unstable features within the compiler.

Reviewer 5: ```suggestion
This does not help with identifying nightlies that a feature is available on or compatible with which will still require a `build.rs`.
This is inherently impossible to do reliably with version numbers as nightly features can change at any time.
It could be done with build probes,
but the Cargo team has previously rejected that idea
([source](https://github.com/rust-lang/cargo/issues/11244#issuecomment-2326780810)).
Whether build probes or nightly version checks,
auto-enabling nightly features
(rather than having users opt-in)
runs counter to the spirit of nightly (works like stable except where you opt-in).
Incorrect checks (i.e., not using a build probe, or a faulty build prove) have broken many crates' nightly builds in the past
and have even caused friction in changing unstable features within the compiler.
```

Author: This is opening too much of a can of worms that is not actually relevant to this RFC and I've now reverted the text and clarified it some with https://github.com/rust-lang/rfcs/pull/3857#discussion_r2350124089.

Possibly there could be a clippy lint specifically about `rust = "<something>"`.
Alternatively, we could try to find a way to structure `--check-cfg` to allow the person defining the `cfg` to decide whether it can be used with `=` or not.
One way of doing this is by allowing the `check-cfg` `since` predicate outside of the `values` predicate,
meaning it works with the `cfg` `since` predicate and not the `=` operator.
Another way would be for the `check-cfg` `since` predicate to never work with `=` but to instead
allow operators inside of the `cfg` `since` predicate, e.g. `#[cfg(since(rust, "=1.95.0"))]`.
However, with the rename of the predicate from `version` to `since`, operators don't fit in as easily.

`--check-cfg` will cause the following to warn:
```rust
fn is_stderr_terminal() -> bool {
    #[cfg(rust)]
    #[cfg(since(rust, "1.70"))]
    use std::io::IsTerminal as _;
    #[cfg(rust)]
    #[cfg(not(since(rust, "1.70")))]
    use is_terminal::IsTerminal as _;
    #[cfg(not(rust))]
    use is_terminal::IsTerminal as _;

    std::io::stderr().is_terminal()
}
```

> 

Reviewer 4: That's a reason for making `values(since(..))` only apply to the `since()` predicate (in `#[cfg]`), otherwise it becomes ambiguous when `--check-cfg` would warn.

Author: Unsure what this is referring to

Author: I'm not seeing how `values(none())` and `since()` not applying to `values()`
- are related
- causes ambiguity

- Feature Name: `cfg_version`
- Start Date: 2025-09-13
- RFC PR: [rust-lang/rfcs#3857](https://github.com/rust-lang/rfcs/pull/3857)
- Rust Issue: [rust-lang/rust#64796](https://github.com/rust-lang/rust/issues/64796)

# Summary
[summary]: #summary

> Allow Rust-version conditional compilation by adding a built-in `--cfg rust=<version>` and a minimum-version `#[cfg]` predicate.

Author: I tweaked this idea a bit in a07e1c69

#[test]
fn ui() {
    let t = trybuild::TestCases::new();
    t.compile_fail("tests/derive_ui/*.rs");
}
```

### Working around compiler bugs

*(non-motivating)*

Requires: programmatic, vendor version
- Can be approximated by using the language version

At times, a vendor's compiler has  bugs that need to be worked around,
e.g. see [error-chain#101](https://github.com/rust-lang-deprecated/error-chain/issues/101).

### Build information

*(non-motivating)*

Requires: display, vendor version
- Can be approximated by using the language version

Some applications choose to include build information in their verbose-version or `--bugreport`.
This can include the compiler vendor and version used to build the application.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

When using a new language or standard library feature,
circumstances may warrant doing so while maintaining an existing [MSRV](https://doc.rust-lang.org/cargo/reference/rust-version.html),
rather than raising to what the language or standard library feature needs.
This can be accomplished by conditionally compiling the code for that feature.

> For instance, say you have an MSRV of 1.10 and `#[cfg(since)]` feature was available in 1.20, you would have been able to do:

Author: I believe this was addressed in ee1c8cfc875319615eaff34abb20d46d66c8f6a9

```rust

> #[cfg_attr(since(rust, "1.27"), must_use)]

Reviewer 6: Text says 1.20 but the example uses 1.27.

Author: `#[must_use]` however is from 1.27.

Tried to clarify this with ee1c8cfc

> ```rust
> #[cfg_alias(must_use_exists, since(rust, "1.27"))]
>
> #[cfg_attr(must_use_exists, must_use)]
> fn double(x: i32) -> i32 {
>     2 * x
> }
>
> fn main() {
>     double(4);
>     // warning: unused return value of `double` which must be used
>     // ^--- This warning only happens if we are on Rust >= 1.27.
> }
> ```

Now, let's say `#[cfg(since)]` was stabilized in Rust 1.27 or later, you can check for support for it with:
```rust
#[cfg_attr(rust, cfg_attr(since(rust, "1.27"), must_use))]
fn double(x: i32) -> i32 {
    2 * x
}

fn main() {
    double(4);
    // warning: unused return value of `double` which must be used
    // ^--- This warning only happens if we are on Rust >= 1.27.
}
```
However, this would produce an `unexpected_cfgs` lint and you would need to add the following to `Cargo.toml`:
```toml
[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(rust,values(none()))'] }
```

Say you were wanting to test out `#[must_use]` after it got stabilized on nightly to provide feedback and to be ready for when it hits stable,
you would instead use `"1.27.0-0"` to match all pre-release versions of 1.27.0:
```rust

> #[cfg_attr(since(rust, "1.27.0-0"), must_use)]

Reviewer: `-0` is the lowest possible prerelease version, since they are ordered by integers first, and all lettered identifiers are after integers

    // ^--- This warning only happens if we are on Rust >= 1.27.
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## `since` cfg predicate

A `since` cfg predicate will be added to Rust.
As Cargo mirrors Rust's `#[cfg]` syntax, it too will gain this predicate.

The [syntax](https://doc.rust-lang.org/reference/conditional-compilation.html#grammar-ConfigurationPredicate) is:
```
ConfigurationVersion -> `since` `(` IDENTIFIER `,` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`
```

If `IDENTIFIER` has no value or is undefined, this will evaluate to `false`.

If `IDENTIFIER` is not a valid [SemVer](https://semver.org/) value, this will evaluate to `false`.

If the string literal does not conform to the syntax from `<major>` to `<major>.<minor>.<patch>-<pre-release>` where the first three fields must be integers, this will evaluate to `false`.
Note that this excludes support for the `+build` field.

Otherwise, the `IDENTIFIER` will be compared to the string literal according to
[Cargo's `>=` version requirements](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#comparison-requirements).

> For example, `#[cfg(since(rust, "1.90"))]` will be treated as `1.95.2 >= 1.90.0`.

Reviewer 7: > Is there a strong reason to evaluate to false here instead of raising an error?

The reason not to error is forwards compatibility; we have to design the feature such that the compiler of today will accept any `cfg(since(crazy_future_value, "future version"))` we might want in the future.

That said, I agree that we should **warn** for incorrect values. I _think_ this is covered by the `--check-cfg` stuff below, but it'd be useful if the RFC called that out here.

`rust` will be specified as `--check-cfg 'cfg(rust, values(since("1.95.0")))'`
(or whatever version this gets stabilized in).

This will be reported back through `--print=cfg`.

Because this gets reported back in `--print=cfg`,
Cargo will expose `rust` in:
- build scripts as `CARGO_CFG_RUST`
- `[target."cfg()".dependencies]`

## clippy

Clippy has a [`clippy::incompatible_msrv`](https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv) lint
which will fire whenever a standard library item is used with a `#[stable(since)]` newer than `package.rust-version`.
However, it will be perfectly reasonable to use those items when guarded by a `#[cfg(since)]`.

Clippy may wish to:
- Find a way to reduce false positives, e.g. evaluating the `cfg(since)`s that led to the item's usage or disabling the lint within `#[cfg(since)]`
- Suggest `#[cfg(since)]` in the `clippy::incompatible_msrv` diagnostic report (maybe along with offering to bump MSRV as that is a reasonable alternative)

# Drawbacks
[drawbacks]: #drawbacks

People may be using `--cfg rust` already and would be broken by this change.
There are no compatibility concerns with predicate names.

> This does not include a solution for adopting this within `Cargo.toml` without waiting for an MSRV bump.

Author: Conceptually its about stacking `cfg`s, that can be done with `#[cfg(rust)] #[cfg(since(rust, ...))]` or `cfg_attr`.

I did make a tweak in c3867009 but not fully the direction you went.

However, with the rename of the predicate from `version` to `since`, operators don't fit in as easily.
If someone wanted to support equality checks, there wouldn't be a way to support a continuous range of `values()` but would instead have to manually specify each likely potential version.

`--check-cfg` will cause the following to warn:
```rust
fn is_stderr_terminal() -> bool {
    #[cfg(rust)]
    #[cfg(since(rust, "1.70"))]
    use std::io::IsTerminal as _;
    #[cfg(rust)]
    #[cfg(not(since(rust, "1.70")))]
    use is_terminal::IsTerminal as _;
    #[cfg(not(rust))]
    use is_terminal::IsTerminal as _;

    std::io::stderr().is_terminal()
}
```

To allow checking for the presence of `rust`, add the following to your `Cargo.toml`:
```toml
[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(rust,values(none()))'] }
```
Alternatively, we could have the built-in `--check-cfg` for `rust` include `values(none())` but:
- When building on an old version, users will see the warning and will likely want to add it anyways.
- We lose out on `--check-cfg` identifying misused.
  Instead, we may wish to add a dedicated predicate intended for "is set".

> ## `rust` cfg

Author: ... that runs counter to the [Reference](https://doc.rust-lang.org/reference/conditional-compilation.html#r-cfg.predicate.option) and the expectations of the pre-RFC discussion participants.

> ## `rust` cfg

Reviewer 7: The alternative, if we want this to work in crates with a lower MSRV, is to add a `cfg_since_available` or similar `cfg` value by default (as has been discussed at length before for `cfg_version`).

> ## `rust` cfg

Author: > To language lawyer a bit, I don't think the reference specifies this incorrectly here, it [says](https://doc.rust-lang.org/reference/conditional-compilation.html#r-cfg.option-spec):
>
> But nowhere does it state that there is any relationship between configuration option "names" and configuration option "key-value pairs".

I'm not seeing how this addresses my understanding of the Reference.  The Reference defines two states:
- set or unset
- when set, name-only or key-value pair

`#[cfg(foo)]` is recorded as checking "set or unset" without any reference to how it is set.  So my reading of it would be that `#[cfg(foo)]` can be used with key-value pairs.

> Aaaanyhow, it would certainly be useful for this RFC (and https://github.com/rust-lang/rfcs/pull/3750 with cfg!(macos)/...) if such a relationship existed. But whether it makes sense to add, I'm not so sure? Adding it would allow for a lot of weird cfgs, see the table below:

In f1dedc5, I do mention a future possibility for a predicate that checks "set or unset".

> ## `rust` cfg

Author: > Though I'd argue that --cfg=rust --cfg=rust="xyz" is a bad choice, I'd prefer --cfg=cfg_since_available --cfg=rust="xyz", since cfg!(rust) is ambiguous.

I intentionally didn't add `--check-cfg cfg(rust, none())` so that `cfg(rust)` will warn.  Users who don't want to bump MSRV for using `since` and don't want an `unexpected_cfgs` warning on older compiler versions, will already need to muck with `check-cfg`. This is also a bit more of an advanced workflow anyways.  For me, this seems like a decent middle ground.

The main question left is for if/when we had `cfg_value!(rust)`, what would be returned?

> ## `rust` cfg

Author: > The main question left is for if/when we had cfg_value!(rust), what would be returned?

I talked this through in 0a839bd

while letting people on pinned nightlies or bisecting nightlies to set a `-Z` to mark the version as incomplete.

In this RFC, we settled on translating `-nightly` to `-incomplete` because:
- Maintainers can adopt stabilized-on-nightly features with `#[cfg(since(rust, "1.100.0-0"))]` (the lowest pre-release for `1.100.0`), keeping friction low while explicitly acknowledging that the unstable feature may change
- Allows build scripts to experiment with other logic when approximating the vendor version from the language version with less of a chance of needing to invoke `rustc` (e.g. detecting nightly)
- It provides extra context when approximating the vendor version from the language version when populating build information

We called the pre-release `-incomplete` to speak to the relationship to the language version.
Other terms like `partial` could as easily apply.
The term `-nightly` would be more discoverable but would convey more of a relationship to the vendor than the language.

As for differentiating between nightlies,
that corresponds more to the vendor version than the language version,
so we do not include that information.

> ## Alternative designs

Reviewer 8: Could we have a vote on the proposed "since" versus the alternative designs?

I personally like them better.

Author: The "vote" will be done by the owning team for this RFC (T-lang).  The responsibility for the rest of us is to provide meaningful feedback that can be included for them to review in their evaluation.

Keep in mind though that part of why this RFC came about is concerns that came up during the *stabilization* of `rust_version(_)` which makes it unlikely we will pivot in that direction.

- Version

## Other

Python
- Programmatic version: [`sys.version_info`](https://docs.python.org/3/library/sys.html#sys.version_info)
- Vendor display version: [`sys.version`](https://docs.python.org/3/library/sys.html#sys.version)
- [Dependency specifiers](https://packaging.python.org/en/latest/specifications/dependency-specifiers/)
  - e.g. `requests [security,tests] >= 2.8.1, == 2.8.* ; python_version < "2.7"`

C++
- Numeric value representing the version of the C++ standard: [`__cplusplus`](https://en.cppreference.com/w/cpp/preprocessor/replace):

C:
- Implementation-defined value of the C++ standard: [__STDC_VERSION__](https://en.cppreference.com/w/cpp/preprocessor/replace)

Haskell:
- Numeric value representing the vendor's version, e.g. `#if __GLASGOW_HASKELL__ >= 706`

# Unresolved questions
[unresolved-questions]: #unresolved-questions

# Future possibilities
[future-possibilities]: #future-possibilities

- In the future the `--check-cfg` `since()` predicate could make the minimum-version field optional,
  matching all version numbers.

## Relaxing SemVer

Instead of requiring the `IDENTIFIER` in the `check-cfg` `since` predicate to be strictly SemVer `major.minor.patch`,

> we could allow abbreviated forms like `major.minor` or even `major`.

Author: The RFC leaves variable precision as a future possibility but steps through how we can get there

See 940b979146e9d1939e19b25f8d5da941a8ff0a6e, b37fe10

In particular, this is important for
- keeping friction down for packages preparing for stabilized-on-nightly features as their `#[cfg(since)]`s can be inserted and "just work" which can be important for getting feedback quickly while the feature is easier to adapt to feedback that can be gained from these users
  - releasing the package while its in this state puts it at risk to be broken if the feature is changed after stabilization

For RFC 2523, they settled on pre-releases being incomplete,
favoring maintainers to adopt stabilized-on-nightly features immediately
while letting people on pinned nightlies or bisecting nightlies to set a `-Z` to mark the version as incomplete.

In this RFC, we settled on translating `-nightly` to `-incomplete` because:
- Maintainers can adopt stabilized-on-nightly features with `#[cfg(since(rust, "1.100.0-0"))]` (the lowest pre-release for `1.100.0`), keeping friction low while explicitly acknowledging that the unstable feature may change
- Allows build scripts to experiment with other logic when approximating the vendor version from the language version with less of a chance of needing to invoke `rustc` (e.g. detecting nightly)

> - It provides extra context when approximating the vendor version from the language version when populating build information

Author: @dtolnay considering you were a major driver for where things landed for the previous work on how `rust_version` treated nightlies, I'd be interested in your thoughts on this alternative take that this new RFC has.

}
```
However, this would produce an `unexpected_cfgs` lint and you would need to add the following to `Cargo.toml`:
```toml
[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(rust,values(none()))'] }
```

Say you were wanting to test out `#[must_use]` after it got stabilized on nightly to provide feedback and to be ready for when it hits stable,
you would instead use `"1.27.0-0"` to match all pre-release versions of 1.27.0:
```rust
#[cfg_attr(since(rust, "1.27.0-0"), must_use)]
fn double(x: i32) -> i32 {
    2 * x
}

fn main() {
    double(4);
    // warning: unused return value of `double` which must be used
    // ^--- This warning only happens if we are on Rust >= 1.27.
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## `since` cfg predicate

A `since` cfg predicate will be added to Rust.
As Cargo mirrors Rust's `#[cfg]` syntax, it too will gain this predicate.

The [syntax](https://doc.rust-lang.org/reference/conditional-compilation.html#grammar-ConfigurationPredicate) is:
```
ConfigurationSince -> `since` `(` IDENTIFIER `,` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`
```

When evaluating `since`,

> 1. If the string literal does not conform to the syntax from `<major>` to `<major>.<minor>.<patch>-<pre-release>` where the first three fields must be integers, this will evaluate to `false`.<br>

Reviewer 4: Another reason for making it a error would be to avoid `not(since(foo, "<invalid>"))` always returning true.

> 1. If the string literal does not conform to the syntax from `<major>` to `<major>.<minor>.<patch>-<pre-release>` where the first three fields must be integers, this will evaluate to `false`.<br>

Author: 940b979 turned them into errors and 75c486b4 added the `not` rationale.

   *(for warning on always-false checks, see `--check-cfg`)*
2. If `IDENTIFIER` is unset, this will evaluate to `false`.
3. If any of the following evaluates to `true` for any cfg entry for `IDENTIFIER`, `since` will evaluate to `true`, otherwise `false`.
    1. If `IDENTIFIER` is name-only, this entry will evaluate to `false`.
    2. If `IDENTIFIER`'s value is not a valid [SemVer](https://semver.org/) value, the compiler will error
       Note that this excludes support for the `+build` field.
    3. Otherwise, the `IDENTIFIER`s value will be compared to the string literal according to
[Cargo's `>=` version requirements](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#comparison-requirements).
       For example, `#[cfg(since(rust, "1.90"))]` will be treated as `1.95.2 >= 1.90.0`.

Examples:
- `cfg(since(unset_name, "1.0.0"))` will be false
- `--cfg name_only` and `cfg(since(name_only, "1.0.0"))` will be false
- `--cfg foo="bird"` and `cfg(since(name_only, "1.0.0"))` will be a compiler error
- `--cfg foo="1.1.0"` and `cfg(since(foo, "bird"))` will be a compiler error

> - `--cfg foo="1.1.0"` and `cfg(since(foo, "1.2.0"))` will be true

Reviewer 9: Is this backwards? I would expect `--cfg foo="1.1.0"` and `cfg(since(foo, "1.2.0"))` to be false and the following line to be true. If not, this feature is very confusing.

    use is_terminal::IsTerminal as _;

    std::io::stderr().is_terminal()
}
```

This supersedes the `cfg_version` subset of [RFC 2523](https://rust-lang.github.io/rfcs/2523-cfg-path-version.html).

# Motivation
[motivation]: #motivation

These problems mostly have solutions today through third-party crates or other patterns but
- That requires knowledge of them when users expect this to "just work" like in other ecosystems
- They slow down build times, requiring at least one build script to be fully built (even in `cargo check`) and then executed.  In one sample "simple webserver", in the dependency tree there were 10 build scripts and 2 proc-macros built for the sake of version detection.
- The workarounds for Rust-version-specific dependencies are less straightforward and difficult to get right.

## Use cases

In considering use cases, there can be different needs.

Specificity:
- Display version: the format is opaque and only intended for showing to users
- Programmatic version: the format is specified and relied on for comparing values

Semantics:
- Language version: versioning of expected / defined behavior, based on the canonical compiler

> - Vendor name/version: identifying the specific compiler being used

Author: From @jmillikin at https://github.com/rust-lang/rfcs/pull/3857#issuecomment-3314425413

> How would this work with non-`rustc` compilers, such as `gccrs`? The current proposal treats the `rustc` release version as a Rust language version, which only really works in languages that have a single implementation (Python, Swift).
> 
> Looking forward into Rust's future slightly, imagine widespread adoption lead to something like today's C/C++ ecosystem. There are the three major toolchains (GCC, Clang, MSVC) plus various vendor compilers. Language version detection uses a vendor-neutral standard (`__STDC_VERSION__`, `__cplusplus`). Compiler-specific feature detection in libraries is uncommon and generally causes a lot of complaints (c.f. gnulib).
> 
> Rust doesn't currently have a versioned language specification, so the available non-vendor identifiers are the edition and the stability annotation feature names. A way to conditionally compile based on those identifiers might be something like this:
> 
> ```rust
> #[cfg(edition >= 2021)]
> mod some_mod_for_2021_edition;
> 
> #[cfg(feature = "fs_try_exists")] // stabilized in rustc v1.81
> use std::fs::exists as fs_exists;
> 
> #[cfg(not(feature = "fs_try_exists"))]
> fn fs_exists(...) { /* open-coded impl of std::fs::exists for older toolchains */ }
> ```
> 
> Regarding conditional compilation based on edition, it's difficult to think of a concrete use case. The edition is part of the `Cargo.toml` library definition, so there's no case where the same source file will be compiled with different editions. In build systems that let the edition vary by detected toolchain (Bazel) the same mechanism can be used to select different source files and `--cfg` flags.
> 
> The stability annotation feature names haven't historically been exposed to stable Rust, so if they or something like them are to be used for compile-time toolchain feature detection then it might be worth considering whether a more structured format is appropriate.

In this RFC, we settled on translating `-nightly` to `-incomplete` because:
- Maintainers can adopt stabilized-on-nightly features with `#[cfg(since(rust, "1.100.0-0"))]` (the lowest pre-release for `1.100.0`), keeping friction low while explicitly acknowledging that the unstable feature may change
  - `-0` is recommended over `-incomplete` or any other value as the exact pre-release value is unspecified.
- Allows build scripts to experiment with other logic when approximating the vendor version from the language version with less of a chance of needing to invoke `rustc` (e.g. detecting nightly)
- It provides extra context when approximating the vendor version from the language version when populating build information

We called the pre-release `-incomplete` to speak to the relationship to the language version.
Other terms like `partial` could as easily apply.
The term `-nightly` would be more discoverable but would convey more of a relationship to the vendor than the language.

As for differentiating between nightlies,
that corresponds more to the vendor version than the language version,
so we do not include that information.

> ## Alternative designs

Author: From @kornelski at https://github.com/rust-lang/rfcs/pull/3857#issuecomment-3325182128

> Currently `#[cfg()]` has its own microsyntax, being a language inside a language, with its own design that is quite different from Rust's expressions. There's a proposal to move away from that, and [make it behave more like normal expressions, using `&&`/`||` instead of `all()`/`any()`](https://github.com/rust-lang/rfcs/pull/3796). Syntactically, this will end up allowing everything matching macro `expr` inside `cfg()`.
> 
> I think it would be great to generally move towards using a syntax that's more natural as Rust expressions, and use `>=` and `<` for comparisons.
> 
> ```rust
> #[cfg(rust_version >= "1.99")]
> #[cfg(rust_version < "1.99")]
> ```
> 
> There's a proposal for `#[stable(since = "semver")]` or `#[doc(since = "semver")]` attribute: [rust-lang/rust#74182](https://github.com/rust-lang/rust/issues/74182) and there's `#[deprecated(since = "v")]` already. Use of `>=` operator avoids giving `since` multiple roles.

Author: From @tmandry at https://github.com/rust-lang/rfcs/pull/3857#issuecomment-3325881428

> What strikes me when considering @kornelski's [comment above](https://github.com/rust-lang/rfcs/pull/3857#issuecomment-3325182128) is that we haven't yet explored the idea of introducing a new _type_ of cfg value for versions to the language.
> 
> This RFC introduces a new cfg predicate and a new check-cfg predicate to match. If we instead had typed cfg values we could redefine predicates like `=` to do something more specific for that type, in addition to defining additional predicates like `>=` and `<`.
> 
> The question is how to introduce values of a type. Following Rust, we can introduce them as `version("1.85")`. So on the command line you would say `--cfg rust=version("1.85")`, and in code you could write
> 
> ```rust
> #[cfg(rust >= version("1.85")]
> #[cfg(rust < version("1.85")]
> ```
> 
> Again, following Rust, we could choose to allow comparison between a version on the left and a string literal on the right. Probably I would defer this for now, but could be convinced either way.
> 
> Beyond the syntactic improvements, this gives us some extra flexibility. For example:
> 
>     * I can write `--check-cfg cfg(rust >= version("1.85"))` and this defines the set of operators that are allowed with that value.
> 
>     * We can choose to enforce that all values of a cfg are the same type. Or in the case of versions, we can enforce that you only have one value to begin with.
> 
>     * As a future possibility we could have something like `cratename = version("1.100+metadata")` check for an exact match to `+metadata` on the LHS, while a more general requirement like `cratename = version("1.100")` would ignore any metadata that happens to exist on the LHS. Supporting this kind of comparison in the current RFC is less straightforward.
> 
> 
> I still support the current RFC; I think it's better than anything we've considered before and it gets the details right. But this other direction is worth considering as a possibility. Using typed comparisons instead of operators on strings feels more Rusty to me.
> 
> We've already reserved space for cfg types with rustc's `--cfg foo="bar"` syntax that requires the double quotes, so we might as well make use of it. And of course, another future possibility is to add integer configs.
> 
> > Currently `#[cfg()]` has its own microsyntax, being a language inside a language, with its own design that is quite different from Rust's expressions. There's a proposal to move away from that, and [make it behave more like normal expressions, using `&&`/`||` instead of `all()`/`any()`](https://github.com/rust-lang/rfcs/pull/3796). Syntactically, this will end up allowing everything matching macro `expr` inside `cfg()`.
> 
> For the record, I'm very much open to this direction. I don't think it cuts strongly against a comparison operator like `since()`, since that already looks like a function call.

Author: From @traviscross at https://github.com/rust-lang/rfcs/pull/3857#issuecomment-3326647794

> > If we instead had typed cfg values we could...
> 
> What I'd really like eventually is to not have a DSL here at all and just use Rust within cfg attributes. We already have a Rust interpreter on hand -- consteval. Obviously nothing from the crate itself would be in scope.
> 
> Clearly I don't think `cfg_version` should wait for that. But when we start talking about typed cfg values, generic predicates, and, presumably therefore, type inference and type checking for cfg, it does make me wonder whether using Rust here and consteval might not be more straightforward after all.

Author: I've covered binary operators and types in 674ec04db98d48315c98dbe3cbdb009425017498.  This greatly enlarges the scope of this RFC which has me worried about this being available in a reasonable time frame and, if we go down this route, I wonder if we should have a short-term solution, like `rust_version("1.95")`.

From @traviscross 

> What I'd really like eventually is to not have a DSL here at all and just use Rust within cfg attributes. We already have a Rust interpreter on hand -- consteval. Obviously nothing from the crate itself would be in scope.

Keep in mind that Cargo also parses and evaluates the syntax.

We'd also have to figure out how concepts like name-only cfg's translates to "rust".

does not address questions around binary operators,
requiring us to work it out.
For example, are the sides of the operator fully swappable?
If we define all comparisons, would `==` be different than `=`?
How will these operators work in the presence of multiple values or a name-only cfg?

Would we allow implicit coercion so you can skip the `version` inside of `cfg`, like `cfg(rust >= "1.95")`?
I would suggest not because this would make it harder to catch bugs where
- The `--cfg` is not a version but you thought it was
- The `--cfg` should be a version but `version()` was left off

Currently, check-cfg does not apply at all to `--cfg` because it is commonly used with `RUSTFLAGS` which
are applied to all packages and would warn that an unknown `IDENTIFIER` is in use for packages that don't care.
We could still skip checking for unknown `IDENTIFIER`s and instead warn on misuse of `IDENTIFIER`s which would increase the chance of catching a mistake (unless a person duplicated there `--cfg` mistake with `--check-cfg`.

Another is how to handle check-cfg.
The proposed syntax is a binary operator but there is no left-hand side in check-cfg.

> Would we accept `cfg(rust, values(>="1.95"))`?

Author: Recorded this in 79f50aab

- users may stumble and be frustrated with missing features from cargo (do we include all unary and binary operators?)
- behavior differences with Cargo may be needed due to different use cases but could lead to user bugs and frustration as it might not match what users are familiar with

If we nest `since` inside `version`,

> - If there is a concern with boundary with `since` conditions that aren't alleviated by the discussion else where,

Reviewer: "elsewhere" is a compound word
```suggestion
- If there is a concern with boundary with `since` conditions that aren't alleviated by the discussion elsewhere,
```

  then this isn't helped because we are still using `since`

> - Its not clear how a user is expected to reason about this (i.e. how do we teach this?)

Reviewer: grammar: "it is"
```suggestion
- It's not clear how a user is expected to reason about this (i.e. how do we teach this?)
```
